src/Middleware/OutputCaching/src/Memory/MemoryOutputCacheStore.cs

```diff
@@ -26,12 +30,16 @@ public ValueTask EvictByTagAsync(string tag, CancellationToken cancellationToken
         {
             if (_taggedEntries.TryGetValue(tag, out var keys))
             {
-                foreach (var key in keys)
+                if (keys != null && keys.Count > 0)
                 {
-                    _cache.Remove(key);
-                }
+                    // Clone the tags as the collection might be updated by the post eviction callback
```

Comment:
Ah, the `_cache.Remove(key)` could in theory run the eviction callback inline if you remove the last item.

I think you could technically write code to handle inline updates:
```csharp
var i = keys.Count;
while (i > 0)
{
    var oldCount = keys.Count;
    foreach (var key in keys)
    {
        _cache.Remove(key);
        i--;
        if (oldCount != keys.Count)
        {
            // eviction callback ran inline, we need to restart the loop to avoid
            // "collection modified while iterating" errors
            break;
        }
    }
}
```

Reply:
Do you think it's worth it? In terms of memory it's just the collection of pointers, the strings are reused. And if all the callbacks are run inline, it might even be worse to create iterators for each entry.

Reply:
Iterators on HashSet are lightweight structs